// AydleEngine/SimpleGameLoop.js
class SimpleGameLoop{
    constructor(){
        this.updateCallback = undefined;
        this.startCallback = undefined;
        this.hasStarted = false;
        this.startTime = 0
        this.deltaTime = 0;
    }

    run(){
        if(this.updateCallback !== undefined){
            console.log('this is not undefined wdf');
            this.hasStarted = true;
            this.startCallback();
            this.startTime = Date.now();
            //process.nextTick(this.test);
            this.tick();
        }
    }

    tick(){
        // store 'this' for process to reference
        let self = this;
        self.updateCallback();
        //return delta in seconds
        this.deltaTime = (Date.now() - this.startTime) / 1000;
        this.startTime = Date.now();

        process.nextTick(function(){self.tick()});
    }

    setUpdate(callback){
        this.updateCallback = callback;
    }

    setStart(callback){
        this.startCallback = callback;
    }
}

module.exports = SimpleGameLoop;
