// server/GameUtils.js
function GameUtils(){}

const cp = require('child_process');
const GAME_PATH = 'node_modules/games/';
let gameList = require('games/games.json');


GameUtils.GetGameList = function()
{
    return gameList;
}

GameUtils.StartGameInstance = function(gameID,socket,io)
{
    let game = GetGameInfo(gameID);
    let moduleName;

    if(game !== undefined)
    {
        moduleName = game.moduleName;
    
        if(socket !== undefined && io !== undefined)
        {
            if(socket.currentRoom.currentGame === undefined)
            {
                //let child = cp.fork(GAME_PATH + moduleName, [JSON.stringify(data)]);
                let child = cp.fork(GAME_PATH + moduleName);
                socket.currentRoom.currentGame = child;
    
                let players = socket.currentRoom.players;
                
                // set up socket handlers for every player
                for(let i = 0; i < players.length; i++)
                {
                    // get the socket object of the player
                    let socketObj = io.sockets.connected[players[i].socketID];
                    let data = {
                        socket : socketObj,
                        io : io
                    }
                    child.send(data);
                }

                io.in(socket.currentRoom.code).emit('onStartGame', {
                    routeName: moduleName,
                    id: gameID,
                    assets: GetGameAssets(gameID)
                })

                child.send({isRunning : true});
            }
        }
    }
}

function GetGameInfo(gameID)
{
    for(let i = 0; i < gameList.length; i++)
    {
        if(gameID === gameList[i].gameID)
        {
            return gameList[i];
        }
    }

    return undefined;
}

function GetGameAssets()
{
    let gameInfo = GetGameInfo();
    
    if(gameInfo !== undefined)
    {
        return gameInfo.assets;
    }
}

module.exports = GameUtils;
