// server/GameUtils.js
function GameUtils(){}

const cp = require('child_process');
const GAME_PATH = 'node_modules/games/';
let gamesFile = require('games/games.json');
let gameList = gamesFile.games;

GameUtils.GetGameList = function()
{
    return gameList;
}

GameUtils.StartGameInstance = function(gameID,socket,io)
{
    // for debug purposes ............
    let temp = 10
    if(socket !== undefined)
        temp = socket;
    // for debug purposes ............

    let game = GetGameInfo(gameID);
    let moduleName;

    if(game !== undefined)
    {
        moduleName = game.module;

        let data = {
            socket : socket,
            io : io,
            sample : temp
        };
    
        let child = cp.fork(GAME_PATH + moduleName, [JSON.stringify(data)]);
    }
        

}

function GetGameInfo(gameID)
{
    for(let i = 0; i < gameList.length; i++)
    {
        if(gameID === gameList[i].gameID)
        {
            return gameList[i];
        }
    }

    return undefined;
}

function GetGameAssets()
{
    let gameInfo = GetGameInfo();
    
    if(gameInfo !== undefined)
    {
        return gameInfo.assets;
    }
}

module.exports = GameUtils;
