// server/Room.js
class Room{
    constructor(roomCode){
        this.code = roomCode;
        this.name = '';
        this.users = [];
        this.games = [];
        this.isPlaying = false;
        this.currentGame = undefined;
    }

    addUser(someUser){
        users.push();
    }

    removeUser(someUser){
        
        let index = -1;

        for(let i = 0; i < users.length; i++){
            if(users[i].userID === someUser.userID){
                index = i;
                break;
            }
        }

        if(index !== -1){
            users.splice(index,1);
            this.checkHost();
        }
    }

    checkHost(){
        if(!this.hostExists()){
            this.appointNewHost();
        }
    }

    hostExists(){
        for(let i = 0; i < this.users.length; i++){
            if(!this.users[i].isAway){
                if(this.users[i].isHost === true){
                    return true;
                } 
            }
        }
        return false;
    }

    appointNewHost(){
        for(let i = 0; i < this.users.length; i++){
            if(!this.users[i].isAway){
                this.users[i].isHost = true;
                break;
            }
        }
    }

    getHost(){
        for(let i = 0; i < this.users.length; i++){
            if(!this.users[i].isAway){
                if(this.users[i].isHost === true){
                    return this.users[i];
                }
            }
        }
        return undefined;
    }

    getUserBySocketID(socketID){
        for(let i = 0; i < this.users.length; i++) {
            if(this.users[i].socketID === socketID){
                return this.users[i];
            }
        }    
        return undefined;
    }

    getUserByUserID(userID){
        for(let i = 0; i < this.users.length; i++) {
            if(this.users[i].userID === userID){
                return this.users[i];
            }
        }    
        return undefined;
    }

    onDisconnect(socketID){
        let user = this.getUserBySocketID(socketID);
        if(user !== undefined){
            if(user.isHost){
                user.isHost = false;
                this.appointNewHost();
            }
        }
    }
}
module.exports = Room;
