// server/RoomUtils.js
function RoomUtils(){}


// Creates a game room that provides suppport for adding and removing of players
RoomUtils.CreateRoom = function()
{
    let _players = [];
    let _games = [];

    let createdRoom =
    {
        players : _players,
        games: _games,
        AddPlayer : function(socketID)
        {
            let p = CreatePlayer(socketID);
            _players.push(p);
            this.AssignHost();
        },
        RemovePlayer : function(socketID)
        {
            let index = -1;

            for(let i = 0; i < _players.length; i++)
            {
                if(_players[i].playerID === socketID)
                {
                    index = i;
                    break;
                }
            }

            if(index > -1)
            {
                _players.splice(index,1);
                this.AssignHost();
            }
            
        },
        AssignHost : function()
        {
            if(_players.length > 0)
                _players[0].isHost = true;
        }
        
    }

    return createdRoom;

}

function CreatePlayer(socketID)
{
    let createdPlayer =
    {
        playerID : socketID,
        score : 0,
        isHost : false
    }

    return createdPlayer;
}



module.exports = RoomUtils;
