// server/RoomUtils.js
function RoomUtils(){}


// Creates a game room that provides suppport for adding and removing of players
RoomUtils.CreateRoom = function(roomCode)
{
    let _code = roomCode;
    let _players = [];
    let _games = [];
    let _targetIndex = 0;

    let createdRoom =
    {
        code : _code,
        players : _players,
        games : _games,
        targetIndex : _targetIndex,
        AddPlayer : function(socketID)
        {
            let p = CreatePlayer(socketID);
            _players.push(p);
            this.AssignHost();
        },
        RemovePlayer : function(socketID)
        {
            let index = -1;

            for(let i = 0; i < _players.length; i++)
            {
                if(_players[i].playerID === socketID)
                {
                    index = i;
                    break;
                }
            }

            if(index > -1)
            {
                _players.splice(index,1);
                this.AssignHost();
            }
            
        },
        AssignHost : function()
        {
            if(_players.length > 0)
                _players[0].isHost = true;
        },
        GetCurrentTarget : function()
        {
            if(_players.length > _targetIndex)
            {
                return _players[_targetIndex];
            }
                
        },
        GetNextTarget : function()
        {
            if(_players.length > 0)
            {
                    
                // reset the target back to the first user if target index is out of bounds
                if((_targetIndex + 1) >= _players.length)
                {
                    _targetIndex = 0;
                }
                else
                {
                    _targetIndex++;
                }

                return _players[_targetIndex]
            }
        }
        
    }

    return createdRoom;

}

function CreatePlayer(socketID)
{
    let createdPlayer =
    {
        playerID : socketID,
        score : 0,
        isHost : false
    }

    return createdPlayer;
}



module.exports = RoomUtils;
