// server/Room.js
class Room{
    constructor(roomCode){
        this.code = roomCode;
        this.name = '';
        this.users = [];
        this.games = [];
        this.isPlaying = false;
        this.currentGame = undefined;
    }

    addUser(someUser){
        users.push();
    }

    appointNewHost(){
        for(let i = 0; i < this.users.length; i++){
            if(!this.users[i].isAway){
                this.users[i].isHost = true;
                break;
            }
        }
    }

    getUserBySocketID(socketID){
        for(let i = 0; i < this.users.length; i++) {
            if(this.users[i].socketID === socketID){
                return this.users[i];
            }
        }    
        return undefined;
    }

    getUserByUserID(userID){
        for(let i = 0; i < this.users.length; i++) {
            if(this.users[i].userID === userID){
                return this.users[i];
            }
        }    
        return undefined;
    }

    onDisconnect(socketID){
        let user = this.getUserBySocketID(socketID);
        if(user !== undefined){
            if(user.isHost){
                user.isHost = false;
                this.appointNewHost();
            }
        }
    }
}
module.exports = Room;
