// games/GenericGame.js
var SimpleGameLoop = require('AydleEngine/SimpleGameLoop');

class GenericGame{
    constructor(){
        this.gameLoop = new SimpleGameLoop();
        this.io = undefined;
        this.roomCodeReference = undefined;
        this.gameStateObjectList = [];
    }

    setupGameHandlers(socket,io){
        // override game handlers here
        if(socket !== undefined){
            if(socket.gameStateObject === undefined){
                socket.gameStateObject = {
                    socketID : socket.id
                };

                this.gameStateObjectList.push(socket.gameStateObject);
            }

            if(this.roomCodeReference === undefined){
                this.roomCodeReference = socket.currentRoom.code;
            }
        }

        if(this.io === undefined && io !== undefined){
            this.io = io;
        }
    }

    runGameInstance(){
        this.gameLoop.setStart(()=>this.start());
        this.gameLoop.setUpdate(()=>this.update());
		this.gameLoop.run();
    }

    stopGameInstance(){
        // clear all gameStateObjects from sockets    
        for(let i = 0 ; i < this.gameStateObjectList.length; i++){
            let socketObj = this.getSocketObject(this.gameStateObjectList[i]);
            if(socketObj !== undefined){
                if(socketObj.gameStateObject !== undefined){
                    delete socketObj.gameStateObject;
                }
            }
        }

        process.exit(0);
    }

    getSocketObject(socketID){
        if(io !== undefined){
            return io.sockets.connected[socketID];
        }
    }
    
    start(){}
    update(){}
}

module.exports = GenericGame;
