// games/PassTheBomb.js
var GenericGame = require('games/GenericGame');

class PassTheBomb extends GenericGame{

	constructor(){
		super();
		this.timeLeft = 10;
		this.targetNumber = -1;
	}

	// cant define const in es6 :((
	static get MIN(){
		return 1;
	}

	static get MAX(){
		return 10;
	}

	setupGameHandlers(socket,io){
		super.setupGameHandlers(socket,io);
		socket.on('pass', (time, number) =>{
			if (number === this.targetNumber) {
				let timer = time;
				this.timeLeft = timer;
				this.targetNumber = this.GetRandNumber(PassTheBomb.MIN,PassTheBomb.MAX);
				this.io.in(this.roomCodeReference).emit('PassBomb', socket.currentRoom.GetNextTarget().playerID, timer, this.targetNumber);
			}
		})

		socket.on('generate', () =>{
			socket.emit('newRandomNumber', GetRandNumber(PassTheBomb.MIN,PassTheBomb.MAX));
		});
	}

	start(){
		super.start();
		this.targetNumber = this.GetRandNumber(PassTheBomb.MIN,PassTheBomb.MAX);
		this.io.in(this.roomCodeReference).emit('PassBomb', socket.currentRoom.GetNextTarget().playerID, timeleft, this.targetNumber);
	}

	update(){
		super.update();
		this.timeLeft -= this.gameLoop.deltaTime;
		if (this.timeLeft <= 0) {		
			this.io.in(this.roomCodeReference).emit('explode');
			process.exit(0);
		}
		console.log(this.timeLeft);
	}

	GetRandNumber(max, min){
		return Math.floor(Math.random() * (max - min + 1)) + min;
	}
	
}

module.exports = PassTheBomb;
