// games/StopBullyingMe.js
var GenericGame = require('games/GenericGame');

class StopBullyingMe extends GenericGame{
    constructor(){
        super();

        this.timeLeft = 10;
    }

    setupGameHandlers(socket,io){
        super.setupGameHandlers(socket,io);

        if(socket !== undefined()){

            socket.on('charge',() => {
                let player = this.playerManager.getPlayerBySocketID(socket.id);
                player.gameStateObject.power++;
                player.score++;
            });
        }

    }

    start(){
        super.start();
        this.addGameStateObjectProperty('power', 0);
    }

    update(){
        super.update();
        this.timeLeft -= this.gameLoop.deltaTime;
        if(this.timeLeft <= 0){
            this.sendPower();
            this.stopGameInstance();
        }
    }

    sendPower(){
        for(let i = 0; i < this.playerManager.players.length; i++){
            let player = this.playerManager.players[i];
            let socketRef = this.getSocketObject(player.socketID);
            socketRef.emit('punch', player.gameStateObject.power);
        }
    }
}
module.exports = StopBullyingMe;
